crates/services/consensus_module/poa/src/sync.rs

```diff
@@ -0,0 +1,245 @@
+use std::{
+    sync::Arc,
+    time::Duration,
+};
+
+use fuel_core_types::{
+    fuel_types::BlockHeight,
+    services::block_importer::ImportResult,
+};
+use tokio::{
+    sync::{
+        broadcast,
+        watch,
+    },
+    time::Instant,
+};
+
+use crate::ports::P2pPort;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum SyncState {
+    NotSynced,
+    Synced,
+}
+
+pub struct SyncTask<T> {
+    min_connected_reserved_peers: usize,
+    time_until_synced: Duration,
+    time_left_until_synced: Duration,
+    p2p_port: T,
+    block_rx: broadcast::Receiver<Arc<ImportResult>>,
+    state_sender: watch::Sender<SyncState>,
+    inner_state: InnerSyncState,
+    notify: Arc<tokio::sync::Notify>,
+}
+
+impl<T> SyncTask<T> {
+    pub fn new(
+        p2p_port: T,
+        min_connected_reserved_peers: usize,
+        time_until_synced: Duration,
+        block_rx: broadcast::Receiver<Arc<ImportResult>>,
+        notify: Arc<tokio::sync::Notify>,
+        state_sender: watch::Sender<SyncState>,
+        block_height: BlockHeight,
+    ) -> Self {
+        let inner_state = {
+            match (min_connected_reserved_peers, time_until_synced) {
+                (0, Duration::ZERO) => InnerSyncState::Synced(block_height),
+                (0, _) => InnerSyncState::SufficientPeers(block_height),
+                _ => InnerSyncState::InsufficientPeers(block_height),
+            }
+        };
+
+        Self {
+            min_connected_reserved_peers,
+            time_until_synced,
+            time_left_until_synced: time_until_synced,
+            p2p_port,
+            block_rx,
+            state_sender,
+            inner_state,
+            notify,
+        }
+    }
+}
+
+impl<T> SyncTask<T>
+where
+    T: P2pPort,
+{
+    pub async fn run(&mut self) -> anyhow::Result<()> {
+        if self.inner_state.is_sufficient() {
+            self.sync_when_sufficient().await?;
+        } else {
+            self.sync().await?;
+        }
+
+        Ok(())
+    }
+
+    async fn sync(&mut self) -> anyhow::Result<()> {
+        let mut reserved_peers_count = self.p2p_port.reserved_peers_count();
+
+        tokio::select! {
+            latest_count = reserved_peers_count.recv() => {
+                if let Ok(latest_count) = latest_count {
+                    if self.inner_state.change_state_on_peers_update(latest_count >= self.min_connected_reserved_peers) {
+                        self.state_sender.send(self.inner_state.sync_state())?;
+                    }
+                }
+            }
+            block = self.block_rx.recv() => {
+                if let Ok(block) = block {
+                    if self.inner_state.change_state_on_block(block) {
```

Comment:
You can't change the state, because you are not connected to a sufficient amount of peers.

Reply:
It does not change state per se, it tries to, but in case if the current state is `InnerSyncState::InsufficientPeers(_)` then it will just update the block height if it's bigger, but the state won't change.